Add partition-after and partition-before

tomdl89 opened the pull request:
Every few months I see a question on the clojurians slack about how to partition a collection by splitting _after_ an element matches a condition. All solutions I see are either less performant than the solution in this PR or much less readable. This is also something I have written in some form many times before, and I've seen in a couple of code bases I've worked in.

It seems to complement the `take-upto` and `drop-upto` functions nicely too. I mostly followed the source code for these, and `partition-by` in clojure.core, but used volatile vectors to store the partitions in the transducer form rather than `java.util.ArrayList` so as to make it more cross-platform friendly. All tests pass.

tomdl89 commented:
I gave it the name `partition-upto` for consistency with `take-upto` and `drop-upto` but I think `partition-after` or somesuch may make more sense.

weavejester commented on src/medley/core.cljc:
```diff
@@ -357,6 +357,35 @@
+       (fn
+         ([] (rf))
+         ([result]
+          (let [result (if-some [v (not-empty @v)]
+                         (unreduced (rf result v))
+                         result)]
+            (rf result)))
```
Perhaps remove the `let` so that it's:

```clojure
          (rf (if-some [v (not-empty @v)]
                (unreduced (rf result v))
                result))
```

tomdl89 replied:
Good spot, thanks

weavejester commented:
I think this seems reasonable. I seem to recall I've run into a use-case for this, myself. I do think that `partition-after` might be more accurate.

I'd also like to get some specific use cases mentioned in this PR as well. Some examples of when this would be useful would help justify it.

tomdl89 commented:
Cool, I'll take a look at the comment(s) and listing some use cases. I did have a thought that `partition-after` might (rightfully) imply the need for a `partition-before` which would use `take-while` to generate `run` in the lazy form. Not sure what you think about having both?

weavejester commented:
Having both is reasonable as long as there's no easy existing way to achieve either. I don't think there is, but I'd like to double check to ensure we're not reimplementing something that can be done conventionally.

tomdl89 commented:
There is an easy way to achieve `partition-before` so I suggest we don't add it. `(partition-by (complement pred) coll)` will do it. I've changed the name and fixed the redundant let binding. lmk if you think it needs anything else to be merged 🙏 .

tomdl89 commented:
Sorry for the noise. I've added `partition-before` because I don't think it can be easily done conventionally. Ready to review.

weavejester commented on src/medley/core.cljc:
```diff
@@ -357,6 +357,64 @@
+         ([result x]
+          (if @split?
+            (let [result (rf result @v)]
+              (vreset! v (if (reduced? result) [] [x]))
```
Why is `v` reset when the result is reduced?

tomdl89 replied:
Because the "completing" arity still cares about `v` even when the result is reduced. Note that I asserted this behaviour as the final assertion of each test, which would fail if we don't reset `v` when the result is reduced.

weavejester replied:
Thanks for the explanation.

weavejester commented on src/medley/core.cljc:
```diff
@@ -357,6 +357,64 @@
+   (lazy-seq
+    (when-let [s (seq coll)]
+      (let [run (take-upto pred s)]
+        (cons run (partition-after pred (lazy-seq (drop (count run) s)))))))))
```
Why is the `drop` wrapped in a `lazy-seq`?

weavejester replied:
Hmm, I guess it's to prevent the seq from `drop` being realized immediately.

weavejester commented on src/medley/core.cljc:
```diff
@@ -357,6 +357,64 @@
+  true. Returns a transducer when no collection is provided."
+  ([pred]
+   (fn [rf]
+     (let [v (volatile! [])]
```
Maybe instead of `v`, `part` would be more descriptive?

tomdl89 replied:
Done

weavejester commented on src/medley/core.cljc:
```diff
@@ -357,6 +357,64 @@
+                (unreduced (rf result v))
+                result)))
+         ([result input]
+          (let [dv @v]
```
Instead of `dv`, which often indicates delta-v or a change in value, perhaps `part-val`?

tomdl89 replied:
Done

weavejester commented on src/medley/core.cljc:
```diff
@@ -357,6 +357,64 @@
+          (rf (if-some [v (not-empty @v)]
+                (unreduced (rf result v))
+                result)))
+         ([result input]
```
Instead of `input`, `x` would be more consistent.

tomdl89 replied:
Done

weavejester commented on src/medley/core.cljc:
```diff
@@ -357,6 +357,64 @@
+         (when-let [s (seq coll)]
+           (let [run (or (seq (take-while cp s))
+                         (cons (first s) (take-while cp (rest s))))]
+             (cons run (part (lazy-seq (drop (count run) s))))))))
```
I'm not sure I like the `drop` and `count`, as it means that we need to run through the sequence several times unnecessarily.

tomdl89 replied:
Yeah I see what you mean, but I'm not sure there is an easy way to determine how much to drop without knowing how many we took. fwiw, this is the approach taken by clojure.core https://github.com/clojure/clojure/blob/clojure-1.10.1/src/clj/clojure/core.clj#L7201

weavejester replied:
Let's keep this for now then, since at least we won't be slower than core, and we can investigate ways of making this more performant later.

weavejester commented on src/medley/core.cljc:
```diff
@@ -357,6 +357,64 @@
+  true. Returns a transducer when no collection is provided."
+  ([pred]
+   (fn [rf]
+     (let [part (volatile! [])
```
I notice that in clojure.core an ArrayList is used. Is there a reason for avoiding that here? Is it just ClojureScript compatibility?

tomdl89 replied:
Yes, exactly. I guess we could use reader conditionals to circumvent? Main downside is readability I suppose.

weavejester replied:
I think conditionals are the way to go here. It does make it less readable, but one of Medley's goals is to be as fast or faster than Clojure core where possible.

tomdl89 replied:
That's now implemented with reader conditionals. lmk what you think.

tomdl89 commented:
In response to
> I'd also like to get some specific use cases mentioned in this PR as well.

I grepped for questions on the clojurians slack. There's plenty:

1. 2022-11-17
> For example, let’s say I have the sequence:
'("BEGIN" "a" "b" "c" "d" "END" "BEGIN" "foo" "bar" "END" "BEGIN" "c" "l" "j" "cool" "END")
And I would, ideally, group this into a sequence of three sequences:
'(("BEGIN" "a" "b" "c" "d" "END") ("BEGIN" "foo" "bar" "END") ("BEGIN" "c" "l" "j" "cool" "END"))

Would be solved by either `(partition-after #{"END"} coll)` or `(partition-before #{"BEGIN"} coll)`

2. 2022-03-11
> Dumb question but what would be the Clojure way for transforming a list of words into a list of sentences? E.g. [ “Hello.” “How” “are” “you?” “I” “am” “fine.” ] => [ “Hello.” “How are you?” “I am fine.“] I’m thinking it’ll involve a reduce.

Would be solved by `(map #(clojure.string/join " " %) (m/partition-after #(re-matches #".*[.?!]$" %) coll))`

3. 2022-07-09
> I’ve tried partition-by but groups items together. I’ve a collection of has-map ordered and I’ve to split them when a value of an item is null... I was looking more for `[{:k 1} {:k 2} {:k nil} {:k 3} {:k 4}] =>(({:k 1} {:k 2}) ({:k nil} {:k 3} {:k 4}))` something like `(split-when f coll)`

Would be solved by `(m/partition-before (comp nil? :k) coll)`

4. 2022-06-25
> Is there a way to do something like partition-by but where it creates a new group whenever the value is true, rather than when it changes? I've written a reduce version, but I assume I'm missing a trick with partition-by or another function
```
  (->> [1 3 2 4 1 2 3]
       (reduce
        (fn [acc n]
          (if (odd? n)
            (conj acc [n])
            (update
             acc
             (dec (count acc))
             #((fnil conj []) % n))))
        []))
;; => [[1] [3 2 4] [1 2] [3]]
```

Would be solved by `(m/partition-before odd? coll)`

5. 2021-11-03
> Trying to figure out a little data transformation puzzle:
• I have a list of strings with a start 0 and end 1 marker
• Sometimes I get them split up by the API I’m working with, e.g.  ["0a1" "0b1" "0c" "c1" "0d1"]
• And I want to transform this list so that I get ["0a1" "0b1" "0cc1" "0d1"]

Would be solved by `(map clojure.string/join (m/partition-before #(re-find #"^0" %) coll))`

I hope these examples are sufficient. There's plenty more but I think they make a decent enough case. FWIW, most answers either suggested using spec, or did an explicitly stateful `partition-by` (by using an atom to track need to split).

weavejester commented:
Thanks for researching up those real-world examples. I think they more than justify the addition of these two functions.

weavejester reviewed (changes requested):
Thanks for updating the code to use mutable arrays/arraylists. I have a few suggestions on how to improve the code.

tomdl89 commented on src/medley/core.cljc:
```diff
@@ -357,6 +357,75 @@
+                         (let [v (vec #?(:clj (.toArray part) :cljs (aclone part)))]
+                           #?(:clj (.clear part) :cljs (set! part -length 0))
+                           (unreduced (rf result v))))]
+            (rf result)))
```
Sorry, yes easy to forget as this is what they do in clojure.core. Corrected again.

tomdl89 commented on src/medley/core.cljc:
```diff
@@ -357,6 +357,75 @@
+            (rf result)))
+         ([result x]
+          (if @split?
+            (let [result (rf result (vec #?(:clj (.toArray part) :cljs (aclone part))))]
```
I avoided doing this initially because I assumed it's (very slightly) less performant. I added an extra commit with `:inline` metadata to fix this issue in clj. I guess this could be a macro to benefit cljs too, but I really dislike not being able to use reader-conditionals in cljc macros, so didn't bother.

tomdl89 replied:
Also added a commit renaming the -clear and -append fns to make the mutation clearer.

weavejester commented:
Thanks for the updates! Does inlining the functions lead to a measurable performance difference when benchmarked with something like Criterium?

tomdl89 commented:
At the time, I quickly tested with `time` and noticed a (probably statistically insignificant) benefit, but having just tested with criterium, the non-inlined version is consistently (but still, probably insignificantly) faster. It's a bit confusing, but anyway, I've removed the inlining commit.

weavejester commented:
The JVM's JIT compiler can make performance difficult to predict. What may be happening in this case is that Clojure's inlining may be interfering with the JVM's attempts to optimize, so once the JVM is warmed up its ever so slightly slower. This is why I'm often cautious about performance optimizations done in absence of benchmarking - I've been too often bitten by this in the past.

weavejester reviewed (approved):
This looks nice and readable, and I think it's as optimized as we can reasonably achieve. Thanks for your work on this.

tomdl89 commented:
That's all squashed now. Thanks for all of the really helpful feedback.

tomdl89 commented:
Anything outstanding I can help with?

weavejester commented:
Thank you for your work on this, and apologies for the delay. This all looks good, so I'll merge it in and look into cutting a new release.

tomdl89 commented:
Any plan to announce on clojurians slack #announce ? I can if not.

weavejester commented:
You can if you want, I wasn't going to as it didn't seem like a major enough change, but maybe I'm being too cautious.

borkdude commented:
Related: https://github.com/weavejester/medley/issues/47
